### DISCUSSION TITLE
Await Slow Response Functions

### DISCUSSION THREAD

**Discussion by thickmn** [zauberzeug/nicegui#1541] [Q&A]
### Question

This isn't really reproducible code, my apologies -- 
I hope you don't mind more of a conceptual ask. 

Following a file upload and a button click event, a batch API request kicks off followed by a few database operations. One fetches results from a server, the other inserts into an embedded SQLite db. The API request function is asynchronous but one of the database operations is synchronous. I'm seeing some weird conditions with the order of the database operations getting out of whack. 

Having little asyncio experience, I can't help but feel like I'm writing workarounds by awaiting with a dialog and asyncio.sleep(). Any obvious mistakes that I'm making? Tips on more elegant approaches? 

My attempt at handling  -- 

```py
class MultiBatch:
    def __init__(self) -> None:
        self.df = None
        self.headers = []
        self.selections = {"member_id": None, "terming_npi": None}
        self.process_completed = False
        self.session = Session(sqlite_engine)
        self.initialize_ui()


    async def handle_upload(self, e: events.UploadEventArguments):
        self.df = self.process_uploaded_file(e)
        self.headers = list(self.df.columns)
        self.on_upload_ui()
        

    def process_uploaded_file(self, e):
        return pd.read_csv(BytesIO(e.content.read()), encoding="utf-8")
    

    async def on_submit_click(self, e, submit_dialog):
        asyncio.create_task(self.submit_handler(e.sender, submit_dialog))


    async def perform_tasks(self, submit_dialog) -> None:
        submit_dialog.open()
        try:
            
            await self.set_status_and_run_task("Creating new input...", self.create_new_input)
            provider_data = await self.set_status_and_run_task("Fetching Provider Pool...", fetch_provider_pool)
            await self.set_status_and_run_task("Performing Termed Provider Data Validation", insert_providers, provider_data)
            await self.set_status_and_run_task("Process Complete", lambda: None)
            
        finally:
            self.submit_result.clear()
            submit_dialog.close()

    async def set_status_and_run_task(self, status, task_func, *args):
        self.submit_result.content = status
        
        if iscoroutinefunction(task_func):
            result = await task_func(*args)
        else:
            loop = asyncio.get_event_loop()
            result = await loop.run_in_executor(None, task_func, *args)

        await asyncio.sleep(1)
        self.submit_result.clear()
        return result


    async def submit_handler(self, button: ui.button, submit_dialog: ui.dialog) -> None:
        with self.disable(button):
            try:
                await self.perform_tasks(submit_dialog)
            except Exception as e:
                print(f"Error in submit_handler: {e}")
            finally:
                self.session.close()`

**Comment by rodja** [comment]
> I'm seeing some weird conditions with the order of the database operations getting out of whack.

I'm sorry, but can you rephrase this sentence? I guess this is your question, right? I just don't know what you want to ask.

> I can't help but feel like I'm writing workarounds by awaiting with a dialog and asyncio.sleep()

Generally your code looks ok. You only need the sleep if you really want the user to read the status messages. The operations execute much faster. Still, awaiting them is good pratice. Even 50-100 ms would be a long time to stop the whole event loop.

**Reply by thickmn** [reply]
Late responding here, but the issue was a field in my ORM model causing the SQLite session to rollback during one of the slow-response functions (and subsequently starting the next query function prematurely)
